File: apigateway/src/main/java/com/openat/apigateway/config/ConcurrencyLimitFilter.java
package com.openat.apigateway.config;

import com.openat.apigateway.error.ApiErrorResponseWriter;
import io.micrometer.core.instrument.MeterRegistry;
import io.micrometer.core.instrument.Tags;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.atomic.AtomicLong;
import lombok.extern.slf4j.Slf4j;
import org.springframework.core.Ordered;
import org.springframework.core.annotation.Order;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Component;
import org.springframework.web.server.ServerWebExchange;
import org.springframework.web.server.WebFilter;
import org.springframework.web.server.WebFilterChain;
import reactor.core.publisher.Mono;

/**
 * 동시에 "처리 중"인 요청 수가 상한을 넘으면 처리를 시작하지 않고 즉시 429로 거절한다.
 *
 * <p>MVC/Tomcat이 스레드 풀 상한(운영 기준 50개) 덕분에 부수 효과로 갖던 "동시 처리 요청 수
 * 제한"을, WebFlux/Netty 기반인 이 게이트웨이에서 명시적으로 재현한 것이다. 근거가 되는 실측
 * 결과와 기본값의 산정 근거는 {@link ConcurrencyLimitProperties} 자바독 참고 - 요약하면 운영과
 * 동일한 힙 상한(약 160MB)에서 동시접속 5,000~6,000명에 <b>3회 시행 전부 OOM으로 다운</b>됐고,
 * 이 필터 적용이 그 대응이다.
 *
 * <p><b>왜 "대기"가 아니라 "즉시 거절"인가</b>: MVC는 넘치는 요청을 OS 커널의 accept backlog에
 * 사실상 공짜로 대기시킬 수 있지만, 이벤트 루프에는 그런 공짜 대기 공간이 없다 - 대기시키려면
 * 그 대기 상태 자체가 또 힙을 쓰므로 문제를 미루기만 한다. 서비스 전체가 죽어 모든 사용자가
 * 못 쓰게 되는 것보다, 상한을 넘는 일부만 지연 없이 실패시켜 클라이언트의 재시도/폴백을 유도하는
 * 쪽이 낫다는 판단이다.
 *
 * <p>{@link Ordered#HIGHEST_PRECEDENCE}로 등록해 인증·라우팅보다 먼저 카운트를 확인한다 -
 * 그래야 상한 초과 시 뒤쪽 처리(본문 버퍼링, 다운스트림 프록시 등)로 인한 추가 메모리 사용
 * 자체를 막을 수 있다. {@code RouteExistenceFilter}도 같은 우선순위지만, 이 필터는 카운터만
 * 증감하고 통과시키므로 둘의 상대 순서는 동작에 영향이 없다.
 *
 * <p>액추에이터(k8s probe/Prometheus scrape) 경로는 상한에서 제외한다 - 과부하로 초과 요청을
 * 거절하는 상황에서 헬스체크까지 429가 되면 쿠버네티스가 정상 동작 중인 파드를 죽여버린다.
 *
 * <p>카운터를 SSE({@code /status/stream})와 일반 API로 분리한다(리뷰 지적으로 추가): SSE는
 * 연결 수명 내내 슬롯을 점유하는 성격이라 일반 제어 API(진입/폴링/GIVE_UP 등 decision)와 같은
 * 카운터를 공유하면, SSE 구독자가 상한을 다 채운 순간부터는 GIVE_UP 같은 상태 변경 요청도
 * 429로 거절돼 사용자가 스스로 자리를 반납할 방법조차 없어진다. 별도 카운터로 나눠 SSE 포화가
 * 제어 API 가용성을 잠식하지 않게 한다.
 */
@Slf4j
@Component
@Order(Ordered.HIGHEST_PRECEDENCE)
public class ConcurrencyLimitFilter implements WebFilter {

    private static final String ACTUATOR_PREFIX = "/actuator";
    private static final String SSE_PATH_SUFFIX = "/status/stream";
    // 과부하 시 거절 건마다 WARN을 남기면 로그 자체가 부하 유발원이 된다(리뷰 지적) - 카운터는
    // Micrometer로 매 건 저비용 집계하고, WARN 로그는 이 값마다 하나씩만 "누적 거절 수"와
    // 함께 남긴다. 상한 초과가 시작된 첫 건은 항상 즉시 로깅해 장애 시작 시점을 놓치지 않는다.
    private static final long REJECT_LOG_SAMPLE_RATE = 100L;

    private final int maxInFlight;
    private final int sseMaxInFlight;
    private final ApiErrorResponseWriter responseWriter;
    private final AtomicInteger inFlight = new AtomicInteger(0);
    private final AtomicInteger sseInFlight = new AtomicInteger(0);
    private final AtomicLong rejectedTotal = new AtomicLong(0);
    private final AtomicLong rejectedTotalSse = new AtomicLong(0);

    public ConcurrencyLimitFilter(ConcurrencyLimitProperties properties,
                                  ApiErrorResponseWriter responseWriter,
                                  MeterRegistry meterRegistry) {
        this.maxInFlight = properties.maxInFlightRequests();
        this.sseMaxInFlight = properties.sseMaxInFlightRequests();
        this.responseWriter = responseWriter;
        meterRegistry.gauge("gateway.concurrency.in-flight", Tags.of("kind", "control"), inFlight);
        meterRegistry.gauge("gateway.concurrency.in-flight", Tags.of("kind", "sse"), sseInFlight);
        meterRegistry.gauge("gateway.concurrency.rejected", Tags.of("kind", "control"), rejectedTotal);
        meterRegistry.gauge("gateway.concurrency.rejected", Tags.of("kind", "sse"), rejectedTotalSse);
    }

    @Override
    public Mono<Void> filter(ServerWebExchange exchange, WebFilterChain chain) {
        String path = exchange.getRequest().getPath().value();
        if (path.startsWith(ACTUATOR_PREFIX)) {
            return chain.filter(exchange);
        }

        boolean isSse = path.endsWith(SSE_PATH_SUFFIX);
        AtomicInteger counter = isSse ? sseInFlight : inFlight;
        int limit = isSse ? sseMaxInFlight : maxInFlight;

        int current = counter.incrementAndGet();
        if (current > limit) {
            counter.decrementAndGet();
            logRejected(isSse, current - 1, limit, path);
            return responseWriter.write(
                    exchange,
                    HttpStatus.TOO_MANY_REQUESTS,
                    "GATEWAY_CONCURRENCY_LIMIT_EXCEEDED",
                    "서버가 처리할 수 있는 동시 요청 수를 초과했습니다. 잠시 후 다시 시도해 주세요.");
        }
        return chain.filter(exchange).doFinally(signal -> counter.decrementAndGet());
    }

    private void logRejected(boolean isSse, int currentInFlight, int limit, String path) {
        AtomicLong counter = isSse ? rejectedTotalSse : rejectedTotal;
        long total = counter.incrementAndGet();
        // 첫 건(total==1)은 즉시, 이후로는 표본만 - Micrometer 게이지가 정확한 누적치를 계속
        // 들고 있으므로 로그가 빠뜨린 건도 관측에서 유실되지 않는다.
        if (total == 1L || total % REJECT_LOG_SAMPLE_RATE == 0L) {
            log.warn("[gateway-concurrency-limit] rejected(sampled 1/{}) kind={} inFlight={} max={} path={} rejectedTotal={}",
                    REJECT_LOG_SAMPLE_RATE, isSse ? "sse" : "control", currentInFlight, limit, path, total);
        }
    }
}

File: queue/src/main/kotlin/com/openat/queue/infrastructure/config/ConcurrencyLimitFilter.kt
package com.openat.queue.infrastructure.config

import com.fasterxml.jackson.databind.ObjectMapper
import com.openat.common.error.ErrorResponse
import com.openat.queue.domain.error.QueueErrorCode
import io.micrometer.core.instrument.MeterRegistry
import io.micrometer.core.instrument.Tags
import org.slf4j.LoggerFactory
import org.springframework.core.Ordered
import org.springframework.core.annotation.Order
import org.springframework.core.io.buffer.DataBufferFactory
import org.springframework.http.HttpStatus
import org.springframework.http.MediaType
import org.springframework.stereotype.Component
import org.springframework.web.server.ServerWebExchange
import org.springframework.web.server.WebFilter
import org.springframework.web.server.WebFilterChain
import reactor.core.publisher.Mono
import java.util.concurrent.atomic.AtomicInteger
import java.util.concurrent.atomic.AtomicLong

/**
 * MVC(Tomcat)는 스레드 풀(운영 기준 50개)이 그 이상 몰리는 요청을 처리 시작 전 단계(OS 커널의
 * 가벼운 accept backlog)에 묶어둬서 "동시 처리 요청 수 제한"을 부수 효과로 갖는다 - 몰려도
 * 힙을 거의 안 쓴다. WebFlux/Netty에는 이런 상한이 기본적으로 없다 - 들어오는 연결을 원칙적으로
 * 전부 받아들여 처리를 시작하므로, 아주 작은 힙(운영 파드 기준 약 220MB)에서 동시접속이 수천
 * 단위로 몰리면 각 요청이 물고 있는 처리 상태(채널·버퍼·리액티브 파이프라인)가 쌓여
 * OutOfMemoryError로 이어진다.
 *
 * 실측으로 확인된 문제: t3.large(2vCPU) + 운영과 동일한 힙(220MB) 근사 환경에서 동시접속을
 * 계단식으로 올리자 약 5,000~6,000명 구간에서 3회 반복 전부 `OutOfMemoryError`로 서비스가
 * 다운됐다(같은 조건의 MVC+폴링은 3회 다 에러 0건으로 버팀 - 응답만 느려짐). 정밀 분석 결과,
 * "WebFlux는 스레드를 적게 쓴다"는 이론 자체는 계속 확인됐지만(스레드 수는 부하와 무관하게
 * 낮고 평평), 그건 메모리 축과는 무관한 별개의 자원 이야기였다 - MVC가 메모리 축에서
 * "우연히" 유리했던 것뿐, WebFlux가 메모리 축을 스스로 보호하는 장치는 원래 없었다.
 * (queue/loadtest/three-stage-story.md §4-4 "정직하게 밝히는 한계" 절 참고)
 *
 * 이 필터는 MVC가 부수 효과로 얻던 그 보호를 명시적으로 재현한다 - 동시 처리 중인 요청 수가
 * 상한을 넘으면 처리를 아예 시작하지 않고 즉시 429로 거절한다. 서비스 전체가 죽어서 모든
 * 사용자가 못 쓰게 되는 것보다, 상한을 넘는 일부 요청만 지연 없이 실패시키는 쪽이 낫다는
 * 판단 - MVC가 넘치는 요청을 accept backlog에 "대기"시키는 것과 달리 이 필터는 "즉시 거절"을
 * 택했다. WebFlux에는 스레드처럼 공짜로 쓸 수 있는 대기 공간이 없고(대기시키려면 그 자체가
 * 또 메모리를 쓴다), 즉시 거절이 재시도/폴백을 유도하기에 더 명확한 신호이기 때문이다.
 *
 * `Ordered.HIGHEST_PRECEDENCE`로 등록해 인증·비즈니스 로직보다 먼저 카운트를 확인한다 -
 * 그래야 상한을 넘었을 때 뒤쪽 처리(Redis 호출 등)로 인한 추가 메모리 사용 자체를 막는다.
 *
 * 액추에이터 경로는 카운트에서 제외한다 - SSE는 연결 수명 내내 in-flight 슬롯을 점유하므로
 * 상한 근처에서는 이 필터가 사실상 "동시 SSE 구독자 상한"으로 작동한다. 제외하지 않으면
 * 부하가 상한을 넘겼을 때 liveness/readiness 프로브까지 429로 거절돼 k8s가 정상 동작 중인
 * 파드를 재시작시킨다([SimulatedIoLatencyFilter]와 동일한 이유로 동일하게 예외 처리).
 *
 * 카운터를 SSE와 일반 API로 분리한다(리뷰 지적으로 추가): SSE(`/status/stream`)는 연결
 * 수명 내내 슬롯을 점유하는 성격이라 일반 제어 API(진입/폴링/GIVE_UP 등 decision)와 같은
 * 카운터를 공유하면, SSE 구독자가 상한을 다 채운 순간부터는 GIVE_UP 같은 상태 변경 요청도
 * 429로 거절돼 사용자가 스스로 자리를 반납할 방법조차 없어진다. 별도 카운터로 나눠 SSE
 * 포화가 제어 API 가용성을 잠식하지 않게 한다.
 */
@Component
@Order(Ordered.HIGHEST_PRECEDENCE)
class ConcurrencyLimitFilter(
    queueProperties: QueueProperties,
    private val objectMapper: ObjectMapper,
    meterRegistry: MeterRegistry,
) : WebFilter {

    private val log = LoggerFactory.getLogger(ConcurrencyLimitFilter::class.java)
    private val maxInFlight = queueProperties.concurrency.maxInFlightRequests
    private val sseMaxInFlight = queueProperties.concurrency.sseMaxInFlightRequests
    private val inFlight = AtomicInteger(0)
    private val sseInFlight = AtomicInteger(0)

    // 과부하 시 거절 건마다 WARN을 남기면 로그 자체가 부하 유발원이 된다(리뷰 지적) - 카운터는
    // Micrometer로 매 건 저비용 집계하고, WARN 로그는 REJECT_LOG_SAMPLE_RATE건마다 하나씩만
    // "누적 거절 수"와 함께 남긴다. 상한 초과가 시작된 첫 건은 항상 즉시 로깅해 장애 시작
    // 시점을 놓치지 않는다.
    private val rejectedTotal = AtomicLong(0)
    private val rejectedTotalSse = AtomicLong(0)

    init {
        meterRegistry.gauge("queue.concurrency.in-flight", Tags.of("kind", "control"), inFlight)
        meterRegistry.gauge("queue.concurrency.in-flight", Tags.of("kind", "sse"), sseInFlight)
        meterRegistry.gauge("queue.concurrency.rejected", Tags.of("kind", "control"), rejectedTotal)
        meterRegistry.gauge("queue.concurrency.rejected", Tags.of("kind", "sse"), rejectedTotalSse)
    }

    override fun filter(exchange: ServerWebExchange, chain: WebFilterChain): Mono<Void> {
        val path = exchange.request.path.value()
        if (path.startsWith("/actuator")) {
            return chain.filter(exchange)
        }

        val isSse = path.endsWith("/status/stream")
        val counter = if (isSse) sseInFlight else inFlight
        val limit = if (isSse) sseMaxInFlight else maxInFlight

        val current = counter.incrementAndGet()
        if (current > limit) {
            counter.decrementAndGet()
            logRejected(isSse, current - 1, limit, path)
            return reject(exchange)
        }
        return chain.filter(exchange)
            .doFinally { counter.decrementAndGet() }
    }

    private fun logRejected(isSse: Boolean, currentInFlight: Int, limit: Int, path: CharSequence) {
        val total = (if (isSse) rejectedTotalSse else rejectedTotal).incrementAndGet()
        // 첫 건(total==1)은 즉시, 이후로는 표본만 - Micrometer 게이지가 정확한 누적치를 계속
        // 들고 있으므로 로그가 빠뜨린 건도 관측에서 유실되지 않는다.
        if (total == 1L || total % REJECT_LOG_SAMPLE_RATE == 0L) {
            log.warn(
                "[concurrency-limit] rejected(sampled 1/{}) kind={} inFlight={} max={} path={} rejectedTotal={}",
                REJECT_LOG_SAMPLE_RATE, if (isSse) "sse" else "control", currentInFlight, limit, path, total,
            )
        }
    }

    private fun reject(exchange: ServerWebExchange): Mono<Void> {
        val response = exchange.response
        response.statusCode = HttpStatus.TOO_MANY_REQUESTS
        response.headers.contentType = MediaType.APPLICATION_JSON
        val body = ErrorResponse.of(QueueErrorCode.CONCURRENCY_LIMIT_EXCEEDED)
        val bytes = objectMapper.writeValueAsBytes(body)
        val bufferFactory: DataBufferFactory = response.bufferFactory()
        return response.writeWith(Mono.just(bufferFactory.wrap(bytes)))
    }

    private companion object {
        const val REJECT_LOG_SAMPLE_RATE = 100L
    }
}
